File: day14/day14.jl
using Nettle

function md5(msg)
	return hexdigest("md5", msg)
end

salt = "cuanljph"
test = "abc"

function hash2(input)
	for i = 1:2017
		input = hexdigest("md5", input)
	end
	return input
end

function findLetter(hashtext)
	for i = 1:30
		if hashtext[i] == hashtext[i + 1] && hashtext[i + 1] == hashtext[i + 2]
			return hashtext[i]
		end
	end
end

function findLetter1(hashtext)
	return match(r"(.)\1\1", hashtext)
end

function evaluate(i, salt, hashfn = md5)
	hashtext = hashfn(salt * string(i))
	letter = findLetter1(hashtext)
	if letter == nothing
		return false
	else
		for j = i + 1: i + 1000
			hashtext = hashfn(salt * string(j))
			if contains(hashtext, repeat(string(letter), 5))
				return true
			end
		end
		return false
	end
end

function nthkey(salt, hashfn = md5, n = 64)
	seen = 0
	guess = 0
	while seen < n
		guess += 1
		if evaluate(guess, salt, hashfn)
			seen += 1
		end
	end
	return guess
end

@time println(nthkey("abc"))
@time println(nthkey("cuanljph"))

File: day14/day14.py
import hashlib
import time

salt = 'cuanljph'
test = 'abc'

def hash1(input):
    return hashlib.md5(input.encode()).hexdigest()
    
def hash2(input):
    for i in range(2017):
        input = hash1(input)
    return input

def findLetter(hashtext):
    for i in range(len(hashtext) - 2):
        if hashtext[i] == hashtext[i + 1] and hashtext[i + 1] == hashtext[i + 2]:
            return hashtext[i]


def evaluate(i, salt = 'abc', hashfn = hash1):
    hashtext = hashfn(salt + str(i))
    letter = findLetter(hashtext)
    if letter == None:
        return False
    else:
        for j in range(i + 1, i + 1001):
            hashtext = hashfn(salt + str(j))
            if ''.join([letter] * 5) in hashtext:
                return True
        return False

def nthkey(salt, hashfn = hash1, n = 64):
    seen = 0
    guess = 0
    while seen < n:
        guess += 1
        if evaluate(guess, salt, hashfn):
            seen += 1
    return guess

a = time.time()
print(nthkey('abc'))
print(nthkey(salt))
print(time.time() - a)

# part 1 takes 11.7 seconds

###
a = time.time()
print(nthkey('abc', hash2))
print(nthkey(salt, hash2))
print(time.time() - a)

# part 2 takes hours right now, needs speed up


